torch/distributed/rpc/api.py

```diff
@@ -592,9 +594,15 @@ def remote(to, func, args=None, kwargs=None, timeout=UNSET_RPC_TIMEOUT):
                 **kwargs,
             )
         else:
-            (pickled_python_udf, tensors) = _default_pickler.serialize(
-                PythonUDF(func, args, kwargs)
-            )
+            arg_cache_key = (func, args, str(kwargs))
+            if arg_cache_key in ARG_CACHE:
+                (pickled_python_udf, tensors) = ARG_CACHE[arg_cache_key]
+            else:
+                (pickled_python_udf, tensors) = _default_pickler.serialize(
+                    PythonUDF(func, args, kwargs)
+                )
+                ARG_CACHE[arg_cache_key] = (pickled_python_udf, tensors)
```

Comment:
This cache assumes that `func` is idempotent, which might not be the case. For example, `func` could modify some global state and its return value might change even with same args. Example:

```
def foo(inp):
   global counter
   counter += 1
   return inp + counter
```

Reply:
@pritamdamania87 I don't believe we actually run `func` anywhere here though? We are just caching its representation, it will still be run on the remote end, as we are not caching the function results.

Also linking  RFC - https://github.com/pytorch/pytorch/issues/49911

Reply:
Aah nvm, my bad I didn't read the code correctly and thought we were caching results.